File: NetDeck/CardImagePopup.m
//
//  CardImagePopup.m
//  Net Deck
//
//  Created by Gereon Steffens on 11.01.14.
//  Copyright © 2016 Gereon Steffens. All rights reserved.
//

#import "CardImagePopup.h"
#import "CardImageCell.h"

@interface CardImagePopup ()

@property CardCounter* cc;
@property Deck* deck;
@property NSUInteger prevCount;

@end

@implementation CardImagePopup

static CardImagePopup* popover;

+(CardImagePopup*) showForCard:(CardCounter *)cc inDeck:(Deck*)deck fromRect:(CGRect)rect inViewController:(UIViewController*)vc subView:(UIView*)view direction:(UIPopoverArrowDirection)direction
{
    popover = [[CardImagePopup alloc] initWithCard:cc andDeck:deck];
    
    popover.modalPresentationStyle = UIModalPresentationPopover;
    popover.popoverPresentationController.sourceRect = rect;
    popover.popoverPresentationController.sourceView = view;
    popover.popoverPresentationController.permittedArrowDirections = direction;
    
    popover.preferredContentSize = popover.view.frame.size;
    
    [vc presentViewController:popover animated:NO completion:nil];

    return popover;
}

+(void) dismiss
{
    if (popover)
    {
        [popover dismissViewControllerAnimated:NO completion:nil];
        popover = nil;
    }
}

- (id)initWithCard:(CardCounter*)cc andDeck:(Deck*)deck
{
    self = [super initWithNibName:@"CardImagePopup" bundle:nil];
    if (self)
    {
        self.cc = cc;
        self.prevCount = cc.count;
        self.deck = deck;
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.copiesStepper.maximumValue = self.deck.isDraft ? 100 : self.cc.card.maxPerDeck;
    self.copiesStepper.value = self.cc.count;
    self.copiesLabel.text = [NSString stringWithFormat:@"×%lu", (unsigned long)self.cc.count];
    self.nameLabel.text = self.cc.card.name;
    
    // auto-scaling of fonts does not work in multiline labels. DIY :(
    int maxDesiredFontSize = 14;
    int minFontSize = 8;
    CGFloat labelWidth = self.nameLabel.frame.size.width;
    CGFloat labelRequiredHeight = self.nameLabel.frame.size.height;
    
    UIFont *font = [UIFont systemFontOfSize:maxDesiredFontSize];
    
    for (int i = maxDesiredFontSize; i > minFontSize; --i)
    {
        font = [font fontWithSize:i];
        
        CGSize bounds = CGSizeMake(labelWidth, MAXFLOAT);
        CGRect textRect = [self.cc.card.name boundingRectWithSize:bounds
                                                          options:NSStringDrawingUsesLineFragmentOrigin
                                                       attributes:@{NSFontAttributeName: font}
                                                          context:nil];

        if (textRect.size.height <= labelRequiredHeight)
        {
            break;
        }
    }
    self.nameLabel.font = font;
}

-(void) copiesChanged:(id)sender
{
    int count = self.copiesStepper.value;
    BOOL delete = count == 0;
    
    NSInteger diff = count - self.prevCount;
    [self.deck addCard:self.cc.card copies:diff];
    self.prevCount = count;
    
    if (delete)
    {
        [CardImagePopup dismiss];
        
        if (self.cc.card.type == NRCardTypeIdentity)
        {
            [[NSNotificationCenter defaultCenter] postNotificationName:Notifications.SELECT_IDENTITY object:self];
            return;
        }
    }
    else
    {
        self.copiesLabel.text = [NSString stringWithFormat:@"×%lu", (unsigned long)count];
    }
    
    self.copiesLabel.textColor = [UIColor blackColor];
    if (!self.deck.isDraft && self.cc.card.owned < count)
    {
        self.copiesLabel.textColor = [UIColor redColor];
    }
    
    [[NSNotificationCenter defaultCenter] postNotificationName:Notifications.DECK_CHANGED object:self];
}

@end

File: NetDeck/CardImagePopup.swift
//
//  CardImagePopup.swift
//  NetDeck
//
//  Created by Gereon Steffens on 01.01.17.
//  Copyright © 2018 Gereon Steffens. All rights reserved.
//

import UIKit

class CardImagePopup: UIViewController {

    @IBOutlet weak var nameLabel: UILabel!
    @IBOutlet weak var copiesLabel: UILabel!
    @IBOutlet weak var copiesStepper: UIStepper!

    private var cc: CardCounter!
    private var deck: Deck!

    static func showFor(cc: CardCounter, inDeck deck: Deck, from rect: CGRect, inViewController vc: UIViewController, subView: UIView, direction: UIPopoverArrowDirection) {
        let popup = CardImagePopup(cc: cc, deck: deck)
        
        popup.modalPresentationStyle = .popover
        popup.popoverPresentationController?.sourceRect = rect
        popup.popoverPresentationController?.sourceView = subView
        popup.popoverPresentationController?.permittedArrowDirections = direction
        
        popup.preferredContentSize = popup.view.frame.size
        
        vc.present(popup, animated: false, completion: nil)
    }
    
    required init(cc: CardCounter, deck: Deck) {
        super.init(nibName: nil, bundle: nil)
        assert(cc.card.type != .identity)
        self.cc = cc
        self.deck = deck
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.copiesStepper.maximumValue = Double(self.deck.isDraft ? 100 : self.cc.card.maxPerDeck)
        self.copiesStepper.value = Double(self.cc.count)
        
        self.copiesLabel.font = UIFont.monospacedDigitSystemFont(ofSize: 15, weight: UIFont.Weight.semibold)
        self.copiesLabel.text = String(format: "×%lu", self.cc.count)
        self.nameLabel.text = self.cc.card.name
    }
    
    @IBAction func copiesChanged(_ stepper: UIStepper) {
        let count = Int(stepper.value)
        let diff = count - self.cc.count
        self.deck.addCard(self.cc.card, copies: diff)
        self.cc.count = count
        
        self.copiesLabel.text = String(format: "×%lu", count)
        self.copiesLabel.textColor = .black
        if !self.deck.isDraft && count > self.cc.card.owned {
            self.copiesLabel.textColor = .red
        }
        
        NotificationCenter.default.post(name: Notifications.deckChanged, object: self)
        
        if count == 0 {
            self.dismiss(animated: false, completion: nil)
        }
    }
}
